go/vt/vttablet/tabletmanager/tm_init.go

```diff
@@ -468,12 +469,19 @@ func (tm *TabletManager) checkMastership(ctx context.Context, si *topo.ShardInfo
 			})
 		case err == nil:
 			if oldTablet.Type == topodatapb.TabletType_MASTER {
+				log.Infof("Shard master alias matches, and existing tablet agrees. Switching to master with tablet's master term start time: %v", oldTablet.MasterTermStartTime)
 				// We're marked as master in the shard record,
 				// and our existing tablet record agrees.
 				tm.tmState.UpdateTablet(func(tablet *topodatapb.Tablet) {
 					tablet.Type = topodatapb.TabletType_MASTER
 					tablet.MasterTermStartTime = oldTablet.MasterTermStartTime
 				})
+			} else {
+				log.Warningf("Shard master alias matches, but existing tablet is not master. Switching to master with the shard's master term start time: %v", oldTablet.MasterTermStartTime)
+				tm.tmState.UpdateTablet(func(tablet *topodatapb.Tablet) {
```

Comment:
The reason we required the tablet record to match too is because the shard record is updated asynchronously and we didn't want to forget a signal that was persisted in the tablet record (see the comment on line 453).

For example:

1. A new master takes over with a higher timestamp, but has not yet asynchronously updated the shard record.
2. The old master is told to become a replica and successfully publishes its tablet record.
3. The old master restarts for some reason, and the shard record still hasn't been updated.
4. The old master reverts from the new state (replica) back into a false master state (with lower timestamp).

I guess we could rely on this being fixed eventually by the shard sync, but it seems kind of ugly to flap back and forth. On the other hand, this converts a different scenario from "requires intervention" to "self-healing" so maybe the potential flapping is justified.

Reply:
> I guess we could rely on this being fixed eventually by the shard sync, but it seems kind of ugly to flap back and forth. On the other hand, this converts a different scenario from "requires intervention" to "self-healing" so maybe the potential flapping is justified.

We're choosing the less of the two bad options. In the case that you listed, we'll flap but will eventually converge. But in the other case where a tablet accidentally overwrites the record as replica, we'll remain in a permanently broken state.